src/internal/reactive-store/index.ts

```diff
@@ -0,0 +1,121 @@
+  unsubscribe(listener: Listener<S>): void;
+}
+
+/**
+ * A pub/sub state management util that registers listeners by selectors.
+ * It comes with React utils that subscribe to state changes and trigger effects or React state updates.
+ *
+ * For simple states, a store can be defined as `ReactiveStore<StateType>`. For more complex states,
+ * it is recommended to create a custom class extending ReactiveStore and providing custom setters,
+ * for example:
+ *   class TableStore extends ReactiveStore<TableState> {
+ *     setVisibleColumns(visibleColumns) {
+ *       this.set((prev) => ({ ...prev, visibleColumns }));
+ *     }
+ *     // ...
+ *   }
+ *
+ * The store instance is usually created once when the component is mounted, which can be achieved with React's
+ * useRef or useMemo utils. To make the store aware of component's properties it is enough to assign them on
+ * every render, unless a state recomputation is required (in which case a useEffect is needed).
+ *   const store = useRef(new TableStore()).current;
+ *   store.totalColumns = props.totalColumns;
+ *
+ * As long as every selector un-subscribes on un-mount (which is the case when `useSelector()` helper is used),
+ * there is no need to do any cleanup on the store itself.
+ */
+export class ReactiveStore<S> implements ReadonlyReactiveStore<S> {
+  private _state: S;
+  private _listeners: [Selector<S, unknown>, Listener<S>][] = [];
+
+  constructor(state: S) {
+    this._state = state;
+  }
+
+  public get(): S {
+    return this._state;
+  }
+
+  public set(cb: (state: S) => S): void {
+    const prevState = this._state;
+    const newState = cb(prevState);
+
+    this._state = newState;
+
+    for (const [selector, listener] of this._listeners) {
+      if (selector(prevState) !== selector(newState)) {
+        listener(newState, prevState);
+      }
+    }
+  }
+
+  public subscribe<R>(selector: Selector<S, R>, listener: Listener<S>): () => void {
+    this._listeners.push([selector, listener]);
+    return () => this.unsubscribe(listener);
+  }
+
+  public unsubscribe(listener: Listener<S>): void {
+    this._listeners = this._listeners.filter(([, storedListener]) => storedListener !== listener);
+  }
+}
+
+/**
+ * Triggers an effect every time the selected store state changes.
+ */
+export function useReaction<S, R>(
+  store: ReadonlyReactiveStore<S>,
+  selector: Selector<S, R>,
+  effect: Listener<R>
+): void {
+  const prevStore = useRef(store);
+  useEffect(
+    () => {
+      if (prevStore.current !== store) {
+        effect(selector(store.get()), selector(prevStore.current.get()));
+        prevStore.current = store;
+      }
+      const unsubscribe = store.subscribe(selector, (next, prev) => effect(selector(next), selector(prev)));
+      return unsubscribe;
+    },
+    // Ignoring selector and effect as they are expected to stay constant.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
This should either be documented for users of the hook, or adjusted so that they do not have to stay constant. This could likely be achieved by using [the `useStableCallback` function](https://github.com/cloudscape-design/component-toolkit/blob/beaa5316188dff51540d7bda5abe1f1bbacfda6b/src/internal/stable-callback/index.ts). The same applies to the `useSelector` hook below